### PR TITLE
RavenDB-17242: Sorting using multiple comparers.

### PR DESCRIPTION
**Pull request by redknightlois**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17242

### Additional description

Implementation of sorting using generic comparers. The current version support up to 9 different comparers. This sorter also support type erasure mechanics in order to avoid having to deal with very complex generic types and the `.Create()` API supports simplified versions with `NullComparer`. The runtime will create the proper code when less than the amount of comparers are used.

### CODE REVIEW

File: test/FastTests/Corax/OrderByMultiSorting.cs [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+            IndexEntries();
+            longList.Sort(CompareAscendingThenDescending);
+            using var searcher = new IndexSearcher(Env);
+            {
+                var match1 = searcher.AllEntries();
+
+                var comparer1 = new AscendingMatchComparer(searcher, Content1, MatchCompareFieldType.Integer);
+                var comparer2 = new DescendingMatchComparer(searcher, Content2, MatchCompareFieldType.Integer);
+               
+                var match = SortingMultiMatch.Create(searcher, match1, comparer1, comparer2);
```
Comment (redknightlois) [inline test/FastTests/Corax/OrderByMultiSorting.cs:77]: This is how to use it @maciejaszyk

File: src/Corax/Queries/SortingMultiMatch.cs [hunk 2]
```diff
@@ -0,0 +1,498 @@
...
+            {
+                read1 = comp1Reader.Read<long>(comparer.FieldId, out var si1);
+                read2 = comp2Reader.Read<long>(comparer.FieldId, out var si2);
+                if (read1 && read2)
+                {
+                    var result = comparer.CompareNumerical<long>(si1, si2);
+                    int nextComparer = comparerIdx + 1;
+                    if (result == 0 && nextComparer < current._totalComparers)
+                    {
+                        return current._compareFuncs[nextComparer](ref current, nextComparer, item1, item2);
```
Comment (redknightlois) [inline src/Corax/Queries/SortingMultiMatch.cs:472]: This effectively generate a sequence of static calls that are number of comparers deep (after the first one). This can be improved upon with some boilerplate code. BUT, unless we prove this is needed with some good profiles showing that this is costly in those cases, I would go with the simple version. cc @ayende

File: test/FastTests/Corax/OrderByMultiSorting.cs (hunk 1 above)
Comment (ayende) [reply test/FastTests/Corax/OrderByMultiSorting.cs:77]: The question is how to do that from an API that doesn't know about this upfront?

File: src/Corax/Queries/SortingMultiMatch.cs [hunk 3]
```diff
@@ -0,0 +1,498 @@
...
+                var read = Get(new IndexEntryReader(matchIndexEntry), fieldId, matches[i], out matchesKeys[i].Value);
+                matchesKeys[i].Key = read ? matches[i] : -matches[i];
+                matchesKeys[i].Entry = matchIndexEntry;
+            }
+
+            // We sort the first batch. That will also mean that we will sort the indexes too. 
+            var sorter = new Sorter<MultiMatchComparer<TComparer1, W>.Item, long, MultiMatchComparer<TComparer1, W>>(comparer);
+            sorter.Sort(matchesKeys[0..totalMatches], matches);
+
+            Span<long> bValues = stackalloc long[matches.Length];
```
Comment (ayende) [inline src/Corax/Queries/SortingMultiMatch.cs:290]: This is probably wrong. we *are* going to get `take int.MaxValue` so you cannot use this.

File: src/Corax/Queries/SortingMultiMatch.cs (hunk 2 above)
Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:472]: I would expect that to be pretty rare, yes.

File: test/FastTests/Corax/OrderByMultiSorting.cs (hunk 1 above)
Comment (redknightlois) [reply test/FastTests/Corax/OrderByMultiSorting.cs:77]: Everything here is constructed in runtime, what do you mean?

File: src/Corax/Queries/SortingMultiMatch.cs (hunk 3 above)
Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: You can. That value comes from the `Fill(...)` the `take` happens after during the algorithm execution.

File: test/FastTests/Corax/OrderByMultiSorting.cs (hunk 1 above)
Comment (ayende) [reply test/FastTests/Corax/OrderByMultiSorting.cs:77]: I'm giving you a `List<string> OrderBy`, what do you do? This is strong typed on the number of sorts, no?

File: src/Corax/Queries/SortingMultiMatch.cs (hunk 3 above)
Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Unless I miss something, if `take` is MaxValue, then `matches` must be at least as long.
If you then also have enough matches, you are going to overflow the stack

File: test/FastTests/Corax/OrderByMultiSorting.cs (hunk 1 above)
Comment (redknightlois) [reply test/FastTests/Corax/OrderByMultiSorting.cs:77]: Yes, I talked with @maciejaszyk about it. Yes that is correct, however we arrived at a tradeoff that would be good enough and will give us a lot of leeway to change the implementation over time. We will relax the requirement for the type to be a `struct` which will give us 3 things:
- In those cases that we cannot know the list of comparer types, we just use `IMatchComparer` instead. 
- We can actually optimize internally all those usages with the front facing API never know that such thing would happen.
- In case of custom comparers we can start with basic interface calls, and then if we need to we just can emit the code to deal with those too if we find edge cases where it matters.

Comment (ayende) [reply test/FastTests/Corax/OrderByMultiSorting.cs:77]: Note that specializing for 2,3 & 4 sort orders should probably suffice for 99.999% of cases, yep

File: src/Corax/Queries/SortingMultiMatch.cs (hunk 3 above)
Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Only if we pass a big enough matches, yeah, we should guard against that anyways...

Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: `from Users where Active = true` is a query that would need that, no?

Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Unless you are willing to pass a multimegabyte buffer to that kind of query. Nope.

Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: ?? That _is_ a query that a user will send, for sure.

Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: I know, but at some point you will have to put a limit... via say 'streaming' or multi-step or something... so that it is bounded by the batch size. The only one that really could need to handle that is sorting. Which we will cross that road when we need to.

Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Not really. This is a real scenario, and we need to give an an answer to that.
Not right away, but when you use `stackalloc` with an unbounded value, we have to ensure that this won't break for the future.

Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Right now, with `Lucene` in this case, it allocates an `int[]` with `Math.Min(take, numOfDocs)` to hold the data.
The `take` is often MaxValue, and the `numOfDocs` depends on the actual size of the index. That can be another limit, but we *are* going to have large indexes, so that is a 1st class worry right now

Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: Yes, that's why we need to think in terms of multi-step when in the face of sorting. So far, it is the only operation that requires you to get all the matches in 1 place. But for that we need to know how many potential matches to have.

Comment (ayende) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: If you need a sorted query over large number of results, you'll just pay the memory cost for that. This is reasonable to expect, as such, we'll just allocate that amount as needed.

Comment (redknightlois) [reply src/Corax/Queries/SortingMultiMatch.cs:290]: That settles the issue then, because I was going to ask for that memory outside of the stack anyways when it goes over 16K elements...
